src/gc.c

```diff
@@ -1953,15 +2005,20 @@ mark: {
 #endif
         jl_taggedvalue_t *o = jl_astaggedvalue(new_obj);
         jl_datatype_t *vt = (jl_datatype_t*)tag;
+        int foreign_alloc = ((void*)o >= sysimg_base && (void*)o < sysimg_end);
         int update_meta = __likely(!meta_updated && !gc_verifying);
         meta_updated = 0;
         // Symbols are always marked
         assert(vt != jl_symbol_type);
         if (vt == jl_simplevector_type) {
             size_t l = jl_svec_len(new_obj);
             jl_value_t **data = jl_svec_data(new_obj);
-            if (update_meta)
-                gc_setmark(ptls, o, bits, l * sizeof(void*) + sizeof(jl_svec_t));
+            if (update_meta) {
+                if (foreign_alloc)
```

Comment:
That's a copy-paste error (weren't supposed to be the same). Sadly, I had implemented this the other way first (as you suggested), then forgot what what I was doing, and changed them all back to this version